internal_api.go

```diff
@@ -228,6 +236,24 @@ func (e *Enforcer) updatePolicies(sec string, ptype string, oldRules [][]string,
 	return ruleUpdated, nil
 }
 
+func (e *Enforcer) updatePoliciesEnhance(sec string, ptype string, oldRule []string, newRules [][]string) (bool, error) {
+	autoSave := e.autoSave
+	e.EnableAutoSave(true)
+
+	oldRule = oldRule[:len(oldRule)-1]
+	policies := e.model.GetFilteredPolicy(sec, ptype, 0, oldRule...)
+	if ok, err := e.removePolicies(sec, ptype, policies); err != nil {
+		return ok, err
+	}
+
+	if ok, err := e.addPolicies(sec, ptype, newRules); err != nil {
+		return ok, err
+	}
+
+	e.EnableAutoSave(autoSave)
+	return true, nil
+}
+
```

Comment:
This implementation do not preserve atomicity. Example: if the call to `e.removePolicies` succeed but the call to `e.addPolicies` fails, we end-up with a partial update / an inconsistent state.

Also, to preserve atomicity between the in-memory policies and the policy storage (adapter), I don't see any other solution than extending the adapter interface (see my previous comment in this PR).

Reply:
@closetool